### ISSUE TITLE
[20.1] compiler/graalunit/LoopTest.java fails with OpenJDK 11u on Linux fastdebug

### ISSUE THREAD

**Issue by christhalinger** [oracle/graal#2449]
This happens with `release/graal-vm/20.1` and the issue is not GC specific (happens with CMS, Parallel, and G1). Also, I can't reproduce the issue on macOS. Running it with OpenJDK 11u:

```
make run-test-only JTREG="VM_OPTIONS=-XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI -XX:+UseJVMCICompiler -XX:+UseG1GC" TEST="compiler/graalunit/LoopTest.java" CONF=fastdebug
```

results in:

```
FAILURES!!!
Tests run: 9,  Failures: 3
```

Here are the three failures (failure 1 and 3 seem to have the same root cause):

1.

```
testSignExtension(org.graalvm.compiler.loop.test.LoopPartialUnrollTest)
java.lang.AssertionError: expected a main loop
	at org.graalvm.compiler.loop.test.LoopPartialUnrollTest.checkMidTierGraph(LoopPartialUnrollTest.java:74)
	at org.graalvm.compiler.core.test.GraalCompilerTest$3.run(GraalCompilerTest.java:307)
	at jdk.internal.vm.compiler@11.0.6.0.4-internal/org.graalvm.compiler.phases.Phase.run(Phase.java:49)
	at jdk.internal.vm.compiler@11.0.6.0.4-internal/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:214)
	at jdk.internal.vm.compiler@11.0.6.0.4-internal/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:147)
	at jdk.internal.vm.compiler@11.0.6.0.4-internal/org.graalvm.compiler.phases.PhaseSuite.run(PhaseSuite.java:209)
	at jdk.internal.vm.compiler@11.0.6.0.4-internal/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:214)
	at jdk.internal.vm.compiler@11.0.6.0.4-internal/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:147)
	at jdk.internal.vm.compiler@11.0.6.0.4-internal/org.graalvm.compiler.core.GraalCompiler.emitFrontEnd(GraalCompiler.java:226)
	at jdk.internal.vm.compiler@11.0.6.0.4-internal/org.graalvm.compiler.core.GraalCompiler.compile(GraalCompiler.java:145)
	at org.graalvm.compiler.core.test.GraalCompilerTest.compile(GraalCompilerTest.java:1112)
	at org.graalvm.compiler.core.test.GraalCompilerTest.getCode(GraalCompilerTest.java:1009)
	at org.graalvm.compiler.core.test.GraalCompilerTest.getCode(GraalCompilerTest.java:950)
	at org.graalvm.compiler.core.test.GraalCompilerTest.executeActual(GraalCompilerTest.java:778)
	at org.graalvm.compiler.core.test.GraalCompilerTest.executeActualCheckDeopt(GraalCompilerTest.java:910)
	at org.graalvm.compiler.core.test.GraalCompilerTest.testAgainstExpected(GraalCompilerTest.java:900)
	at org.graalvm.compiler.core.test.GraalCompilerTest.testAgainstExpected(GraalCompilerTest.java:896)
	at org.graalvm.compiler.core.test.GraalCompilerTest.test(GraalCompilerTest.java:865)
	at org.graalvm.compiler.core.test.GraalCompilerTest.test(GraalCompilerTest.java:834)
	at org.graalvm.compiler.core.test.GraalCompilerTest.test(GraalCompilerTest.java:827)
	at org.graalvm.compiler.loop.test.LoopPartialUnrollTest.testSignExtension(LoopPartialUnrollTest.java:247)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at com.oracle.mxtool.junit.MxJUnitWrapper.runRequest(MxJUnitWrapper.java:339)
	at com.oracle.mxtool.junit.MxJUnitWrapper.main(MxJUnitWrapper.java:220)
```

2. 

```
testLoopCarriedDuplication(org.graalvm.compiler.loop.test.LoopPartialUnrollTest)
java.lang.AssertionError: Graphs do not have the same number of nodes: 33 vs. 29
mismatch in graphs:
========= expected (StructuredGraph:641{HotSpotMethod<LoopPartialUnrollTest.testLoopCarriedReference(int)>}) =========
6 constants: <<<
Constant(0, i32)    (6)
Constant(0.99, f64)    (1)
Constant(1, i32)    (1)
Constant(120, i64)    (1)
Constant(2, i32)    (1)
Constant(Class:org.graalvm.compiler.loop.test.LoopPartialUnrollTest, a -)    (1)
Block B0 * -> B1 
  0|StartNode    (0)
  1|ParameterNode    (1)
  2|OffsetAddressNode    (2)
  3|EndNode    (1)
Block B1 -> B2 B3 
  4|LoopBeginNode    (6)
  5|IntegerLessThanNode    (1)
  6|ConditionalNode    (1)
  7|HotSpotDirectCallTargetNode    (1)
  8|AddNode    (1)
  9|AddNode    (1)
  10|InvokeNode    (2)
  11|IntegerEqualsNode    (1)
  12|IfNode    (0)
Block B2 -> 
  13|LoopExitNode    (0)
  14|ReturnNode    (0)
Block B3 -> B1 
  15|BeginNode    (0)
  16|VolatileReadNode    (2)
  17|AddNode    (2)
  18|AddNode    (1)
  19|VolatileReadNode    (1)
  20|AddNode    (1)
  21|LoopEndNode    (0)

========= actual (StructuredGraph:657{HotSpotMethod<LoopPartialUnrollTest.testLoopCarriedSnippet(int)>}) =========
5 constants: <<<
Constant(0, i32)    (6)
Constant(0.99, f64)    (1)
Constant(1, i32)    (2)
Constant(120, i64)    (1)
Constant(Class:org.graalvm.compiler.loop.test.LoopPartialUnrollTest, a -)    (1)
Block B0 * -> B1 
  0|StartNode    (0)
  1|ParameterNode    (1)
  2|OffsetAddressNode    (1)
  3|EndNode    (1)
Block B1 -> B2 B3 
  4|LoopBeginNode    (6)
  5|IntegerLessThanNode    (1)
  6|ConditionalNode    (1)
  7|HotSpotDirectCallTargetNode    (1)
  8|AddNode    (1)
  9|AddNode    (1)
  10|InvokeNode    (2)
  11|IntegerEqualsNode    (1)
  12|IfNode    (0)
Block B2 -> 
  13|LoopExitNode    (0)
  14|ReturnNode    (0)
Block B3 -> B1 
  15|BeginNode    (0)
  16|VolatileReadNode    (1)
  17|AddNode    (1)
  18|LoopEndNode    (0)
	at org.junit.Assert.fail(Assert.java:88)
	at org.graalvm.compiler.core.test.GraalCompilerTest.assertEquals(GraalCompilerTest.java:440)
	at org.graalvm.compiler.core.test.GraalCompilerTest.assertEquals(GraalCompilerTest.java:415)
	at org.graalvm.compiler.loop.test.LoopPartialUnrollTest.testDuplicateBody(LoopPartialUnrollTest.java:344)
	at org.graalvm.compiler.loop.test.LoopPartialUnrollTest.testLoopCarriedDuplication(LoopPartialUnrollTest.java:103)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at com.oracle.mxtool.junit.MxJUnitWrapper.runRequest(MxJUnitWrapper.java:339)
	at com.oracle.mxtool.junit.MxJUnitWrapper.main(MxJUnitWrapper.java:220)
```

3. 

```
testComplex(org.graalvm.compiler.loop.test.LoopPartialUnrollTest)
java.lang.AssertionError: expected a main loop
	at org.graalvm.compiler.loop.test.LoopPartialUnrollTest.checkMidTierGraph(LoopPartialUnrollTest.java:74)
	at org.graalvm.compiler.core.test.GraalCompilerTest$3.run(GraalCompilerTest.java:307)
	at jdk.internal.vm.compiler@11.0.6.0.4-internal/org.graalvm.compiler.phases.Phase.run(Phase.java:49)
	at jdk.internal.vm.compiler@11.0.6.0.4-internal/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:214)
	at jdk.internal.vm.compiler@11.0.6.0.4-internal/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:147)
	at jdk.internal.vm.compiler@11.0.6.0.4-internal/org.graalvm.compiler.phases.PhaseSuite.run(PhaseSuite.java:209)
	at jdk.internal.vm.compiler@11.0.6.0.4-internal/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:214)
	at jdk.internal.vm.compiler@11.0.6.0.4-internal/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:147)
	at jdk.internal.vm.compiler@11.0.6.0.4-internal/org.graalvm.compiler.core.GraalCompiler.emitFrontEnd(GraalCompiler.java:226)
	at jdk.internal.vm.compiler@11.0.6.0.4-internal/org.graalvm.compiler.core.GraalCompiler.compile(GraalCompiler.java:145)
	at org.graalvm.compiler.core.test.GraalCompilerTest.compile(GraalCompilerTest.java:1112)
	at org.graalvm.compiler.core.test.GraalCompilerTest.getCode(GraalCompilerTest.java:1009)
	at org.graalvm.compiler.core.test.GraalCompilerTest.getCode(GraalCompilerTest.java:950)
	at org.graalvm.compiler.core.test.GraalCompilerTest.executeActual(GraalCompilerTest.java:778)
	at org.graalvm.compiler.core.test.GraalCompilerTest.executeActualCheckDeopt(GraalCompilerTest.java:910)
	at org.graalvm.compiler.core.test.GraalCompilerTest.testAgainstExpected(GraalCompilerTest.java:900)
	at org.graalvm.compiler.core.test.GraalCompilerTest.testAgainstExpected(GraalCompilerTest.java:896)
	at org.graalvm.compiler.core.test.GraalCompilerTest.test(GraalCompilerTest.java:865)
	at org.graalvm.compiler.core.test.GraalCompilerTest.test(GraalCompilerTest.java:834)
	at org.graalvm.compiler.core.test.GraalCompilerTest.test(GraalCompilerTest.java:827)
	at org.graalvm.compiler.loop.test.LoopPartialUnrollTest.testComplex(LoopPartialUnrollTest.java:230)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at com.oracle.mxtool.junit.MxJUnitWrapper.runRequest(MxJUnitWrapper.java:339)
	at com.oracle.mxtool.junit.MxJUnitWrapper.main(MxJUnitWrapper.java:220)
```

**Comment by christhalinger** [comment]
@tkrodriguez who is the loop unrolling expert?

**Comment by tkrodriguez** [comment]
I'm a good place to start.  It appears that it didn't partially unroll the loop.  It seems like we should have a check that actually perform duplicate here https://github.com/oracle/graal/blob/master/compiler/src/org.graalvm.compiler.loop.test/src/org/graalvm/compiler/loop/test/LoopPartialUnrollTest.java#L323.  I'm guessing that for some reason this loop doesn't appear to be counted.  We had a problem in CountedLoopTest where profiling information was screwing up the test under jtreg.  Maybe this test needs a similar fix applied?  https://github.com/oracle/graal/commit/d18b88f6d2e4c88553bbbfd2db2a64efffca1502#diff-79efa3569c04a8f3c2c86592a65dc995

**Comment by christhalinger** [comment]
Do you have any idea why it only fails with fastdebug?

**Comment by tkrodriguez** [comment]
I'm guessing it's related to speed of compilation.  We were never able to reproduce the other failure outside of jtreg and mach5.  In any case this test really shouldn't be paying attention to profiling information.  If you can reproduce it then a BGV dump might tell us for sure why it's failing.  It's clearly not partially unrolled and in this test that would only happen if it wasn't seen as a counted loop.

**Comment by christhalinger** [comment]
If I add `-Dgraal.DumpOnError=true` it passes every time.

**Comment by tkrodriguez** [comment]
That's odd.  Try something like this:
```
diff --git a/src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.loop.test/src/org/graalvm/compiler/loop/test/LoopPartialUnrollTest.java b/src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.loop.test/src/org/graalvm/compiler/loop/test/LoopPartialUnrollTest.java
index 538e9c4b98c..c37f96aae2b 100644
--- a/src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.loop.test/src/org/graalvm/compiler/loop/test/LoopPartialUnrollTest.java
+++ b/src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.loop.test/src/org/graalvm/compiler/loop/test/LoopPartialUnrollTest.java
@@ -24,6 +24,7 @@
  */
 package org.graalvm.compiler.loop.test;
 
+import java.util.Collection;
 import java.util.ListIterator;
 
 import org.graalvm.compiler.api.directives.GraalDirectives;
@@ -319,7 +320,12 @@ public class LoopPartialUnrollTest extends GraalCompilerTest {
             if (partialUnroll) {
                 LoopsData dataCounted = new LoopsData(graph);
                 dataCounted.detectedCountedLoops();
-                for (LoopEx loop : dataCounted.countedLoops()) {
+                Collection<LoopEx> counted = dataCounted.countedLoops();
+                if (counted.size() == 0) {
+                    graph.getDebug().forceDump(graph, "no counted loops");
+                }
+                assertTrue(counted.size() > 0, "expected to find counted loops");
+                for (LoopEx loop : counted) {
                     LoopFragmentInside newSegment = loop.inside().duplicate();
                     newSegment.insertWithinAfter(loop, null);
                 }
```

**Comment by christhalinger** [comment]
Ok I got something. Should I attach the dump here or send it to you directly?

**Comment by tkrodriguez** [comment]
The dump indicated that GraalCompilerTest.branchProbability wasn't inlined so the GraalDirective was never seen and the probability wasn't injected.  We should probably eliminate that override since it relies on the InlineDuringParsingPlugin to work.  It's still unclear why the plugin didn't inline that method though.

**Comment by tkrodriguez** [comment]
@christhalinger did you ever get any insight into the InlineDuringParsingPlugin failed?  The only dynamic pieces of information that control that are the isLinked state which must be true because the caller is a subtype of the declaring class and ResolvedJavaMethod.canBeInlined which could only be stopped by a method being marked not compilable.  Removing the helper makes sense to me but I'm concerned that we're just hiding another issue.

**Comment by christhalinger** [comment]
Sorry, I got distracted and then forgot about it. @usrinivasan will pick this one up.

**Comment by tkrodriguez** [comment]
I think it's clear how to make the problem go away but it's unclear to me how it can even happen.  I'd just like to get clarity on that before pushing a fix for it.

**Comment by tkrodriguez** [comment]
It appears that the use of -XX:+UseJVMCICompiler when running the test can cause the helper to be compiled.  The compilation fails because the injected probability isn't associated with an IfNode and that marks the method as not compilable which then breaks the tests.  So the near term fix is just to remove the helper as I suspected.

**Comment by tkrodriguez** [comment]
This should be resolved by 62564314ddabb1b914dbb51566c1cd11378d1966
